base/compiler/ssair/inlining2.jl

```diff
@@ -0,0 +1,759 @@
+
+        atype_unlimited = argtypes_to_type(atypes)
+
+        # In :invoke, make sure that the arguments we're passing are a subtype of the
+        # signature we're invoking.
+        (invoke_data === nothing || atype_unlimited <: invoke_data.types0) || continue
+
+        # TODO: Bail out here if inlining is disabled
+
+        # Special case inliners for regular functions
+        if late_inline_special_case!(ir, idx, stmt, atypes, f, ft, _topmod(ir.mod))
+            continue
+        end
+
+        # Compute the limited type if necessary
+        if length(atype_unlimited.parameters) - 1 > sv.params.MAX_TUPLETYPE_LEN
+            atype = limit_tuple_type(atype_unlimited, sv.params)
+        else
+            atype = atype_unlimited
+        end
+
+        # Ok, now figure out what method to call
+        if invoke_data === nothing
+            min_valid = UInt[typemin(UInt)]
+            max_valid = UInt[typemax(UInt)]
+            meth = _methods_by_ftype(atype, 1, sv.params.world, min_valid, max_valid)
+            if meth === false || length(meth) != 1
+                maybe_make_invoke!(ir, idx, stmt.typ, atypes, sv, atype_unlimited, false, isapply, nothing)
+                continue
+            end
+            meth = meth[1]::SimpleVector
+            metharg = meth[1]::Type
+            methsp = meth[2]::SimpleVector
+            method = meth[3]::Method
+        else
+            method = invoke_data.entry.func
+            (metharg, methsp) = ccall(:jl_type_intersection_with_env, Any, (Any, Any),
+                                    atype_unlimited, method.sig)::SimpleVector
+            methsp = methsp::SimpleVector
+        end
+
+
+        methsig = method.sig
+        if !(atype <: metharg)
+            maybe_make_invoke!(ir, idx, stmt.typ, atypes, sv, atype_unlimited, isinvoke, isapply, invoke_data)
+            continue
+        end
+
+        # Check whether this call just evaluates to a constant
+        if isa(f, widenconst(ft)) && !isdefined(method, :generator) && method.pure &&
+                isa(stmt.typ, Const) && stmt.typ.actual && is_inlineable_constant(stmt.typ.val)
+            ir[SSAValue(idx)] = quoted(stmt.typ.val)
+            continue
+        end
+
+        # Check that we habe the correct number of arguments
+        na = Int(method.nargs)
+        npassedargs = length(atypes)
+        if na != npassedargs && !(na > 0 && method.isva)
+            # we have a method match only because an earlier
+            # inference step shortened our call args list, even
+            # though we have too many arguments to actually
+            # call this function
+            continue
+        end
+
+        # Bail out if any static parameters are left as TypeVar
+        ok = true
+        for i = 1:length(methsp)
+            isa(methsp[i], TypeVar) && (ok = false; break)
+        end
+        ok || continue
+
+        # Find the linfo for this methods (Generated functions are expanded here if necessary)
```

Comment:
expanding a generated function here would be a bug – this is supposed to be just a cache lookup

Reply:
Ok, I didn't actually change anything here, just added the comment. I guess I confused this function with `code_for_staged`.